File: day1/rust/src/main.rs
use std::fs;

fn get_input_numbers() -> Vec<u64> {
    let input_file = "../input.txt";
    fs::read_to_string(input_file)
        .unwrap()
        .lines()
        .map(|s| s.parse().unwrap())
        .collect()
}

fn get_required_fuel(mass: u64) -> u64 {
    (mass / 3).checked_sub(2).unwrap_or(0)
}

fn get_total_required_fuel(mass: u64) -> u64 {
    let mut required_fuel: u64 = get_required_fuel(mass);
    let mut unaccounted_fuel = required_fuel;
    while unaccounted_fuel > 0 {
        let additional_fuel = get_required_fuel(unaccounted_fuel);
        required_fuel += additional_fuel;
        unaccounted_fuel = additional_fuel;
    }
    required_fuel
}

fn main() {
    let module_masses = get_input_numbers();
    let required_fuel: u64 = module_masses
        .iter()
        .map(|&mass| get_required_fuel(mass))
        .sum();
    println!("Required fuel (part 1): {}", required_fuel);

    let required_fuel: u64 = module_masses
        .iter()
        .map(|&mass| get_total_required_fuel(mass))
        .sum();
    println!("Required fuel (part 2): {}", required_fuel);
}

File: day1/python/main.py
import numpy as np


def get_required_fuel(masses):
    return np.maximum(masses // 3 - 2, 0)


input_file = "../input.txt"

with open(input_file, 'r') as f:
    input_text = f.read()

module_masses = np.array([int(line) for line in input_text.splitlines()])
required_fuel = np.sum(get_required_fuel(module_masses))

print(f"Required fuel (part 1): {required_fuel}")


required_fuel = 0
for module_mass in module_masses:
    required_module_fuel = get_required_fuel(module_mass)
    unaccounted_fuel = required_module_fuel
    while unaccounted_fuel > 0:
        additional_fuel = get_required_fuel(unaccounted_fuel)
        required_module_fuel += additional_fuel
        unaccounted_fuel = additional_fuel
    required_fuel += required_module_fuel

print(f"Required fuel (part 2): {required_fuel}")
